addons/hr_expense/models/product_product.py

```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    standard_price_update_warning_message = fields.Char(compute="_compute_standard_price_update_warning_message", store=False)
+
+    @api.depends('standard_price')
+    def _compute_standard_price_update_warning_message(self):
+        unposted_expenses = self.env['hr.expense'].search([('state', 'in', ['draft', 'reported', 'approved']), ('product_id', 'in', self.ids)])
+        for product in self:
+            product.standard_price_update_warning_message = False
+            if unposted_expenses and not (product.standard_price in [expense.unit_amount if expense.product_id.id == product.id else False for expense in unposted_expenses]+[0.0]):
+                product.standard_price_update_warning_message = \
+                    _("There are unposted expenses linked to this category. Updating the category cost will change expense amounts. Make sure it is what you want to do.")
```

Comment:
if the message is too big you could split the string into two, better to split it than having a "\" imo